trieste/acquisition/function.py

```diff
@@ -366,6 +367,168 @@ def __call__(self, x: TensorType) -> TensorType:
         return expected_improvement * augmentation
 
 
+class MonteCarloExpectedImprovement(SingleModelAcquisitionBuilder):
+    """
+    Builder for a Monte Carlo-based expected improvement function for use with a model where we have
+    to obtain a Monte Carlo estimate of the expected improvement. The "best" value is taken to be
+    the minimum of the posterior mean at observed points.
+    """
+
+    def __init__(self, sample_size: int):
+        """
+        :param sample_size: The number of samples for each batch of points.
+        :raise tf.errors.InvalidArgumentError: If ``sample_size`` is not positive.
+        """
+        tf.debugging.assert_positive(sample_size)
+
+        super().__init__()
+
+        self._sample_size = sample_size
+
+    def __repr__(self) -> str:
+        """"""
+        return f"MonteCarloExpectedImprovement({self._sample_size!r})"
+
+    def prepare_acquisition_function(
+        self,
+        model: ProbabilisticModel,
+        dataset: Optional[Dataset] = None,
+    ) -> AcquisitionFunction:
+        """
+        :param model: The model over the specified ``dataset``. Must have event shape [1].
+        :param dataset: The data from the observer. Must be populated.
+        :return: The estimated *expected improvement* acquisition function.
+        :raise ValueError (or InvalidArgumentError): If ``dataset`` is not populated, ``model``
+            does not have an event shape of [1] or does not have a ``reparam_sample`` method.
+        """
+        get_sampler = getattr(model, "reparam_sampler", None)
+        if not callable(get_sampler):
+            raise ValueError(
+                "The model must have a `reparam_sampler` method, which takes `num_samples` and "
+                "which returns a reparameterization `ModelSampler` to create smooth samples at "
+                "query points."
+            )
+
+        sampler = get_sampler(self._sample_size)
+
+        if not isinstance(sampler, ModelSampler):
+            raise ValueError(
+                "The sampler returned by the model's `reparam_sampler` method must be an instance "
+                "of a `trieste.models.sampler.ModelSampler`."
+            )
+
+        if not isinstance(dataset, Dataset):
+            raise ValueError(
+                f"`dataset` must be a Dataset for {self.__repr__()}, received" f"{type(dataset)}"
+            )
+
+        tf.debugging.assert_positive(len(dataset))
+
+        samples_at_query_points = sampler.sample(dataset.query_points)
```

Comment:
I think this is risky, since for small sample sizes (which would be typical) it may be possible to end up with a case where the expected improvement is zero everywhere due to the different samples being used. @henrymoss do you have any thoughts?